server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -83,6 +83,7 @@
     private String routing;
 
     private BytesReference source;
+    private Map<String, Object> sourceAsMap;
```

Comment:
The timestamp is always going to be in the `@timestamp` field. So maybe instead of caching the entire source just cache the timestamp itself?

Reply:
I think the cached sourceAsMap may be needed elsewhere as well. Time series index will parse routing fields from the source, it can reuse the sourceAsMap.

Reply:
> Time series index will parse routing fields from the source, it can reuse the sourceAsMap.

I rather not keep a map of maps copy of the source around here. Let's focus on routing in a followup change? I think we can both timestamp and routing fields post ingest time without carrying around a bloated map of maps.